KafkaProducer serializer broken when using PublisherBuilder to convert to Message

I noticed when updating the WildFly quickstart that the KafkaProducer gets confused about what the payload is when PublisherBuilder.map() has been used to create a message from a stream of the payloads.

I have modified the SR RM quickstart to showcase the problem here. https://github.com/smallrye/smallrye-reactive-messaging/compare/main...kabir:publisher-with-message-problem?expand=1#diff-db6074d3260024cab0f3ac87e05a6b1f4c5889f07e181960b2d08fbc1ab4d26c (the first commit is me deleting a bunch of stuff to avoid wiring errors, and adjusting the class path to be able to run it)

Essentially I have changed
```
    @Outgoing("prices-out")
    public Multi<Message<Double>> generate() {
        // Build an infinite stream of random prices
        // It emits a price every 8 seconds
        return Multi.createFrom().ticks().every(Duration.ofSeconds(8))
                .map(x -> Message.of(random.nextDouble()));
    }
```
to
```
    @Outgoing("one")
    public Multi<Message<Double>> source() {
        // Build an infinite stream of random prices
        // It emits a price every 8 seconds
        return Multi.createFrom().ticks().every(Duration.ofSeconds(5))
                .map(x -> Message.of(random.nextDouble()));
    }

    @Incoming("one")
    @Outgoing("prices-out")
    public Publisher<Message<Double>> generate(Publisher<Double> messages) {
        return ReactiveStreams.fromPublisher(messages)
                .map(d -> Message.of(d))
                .buildRs();
    }
```
When I do it this way, I get errors as the following
```
rg.apache.kafka.common.errors.SerializationException: Can't convert value of class org.eclipse.microprofile.reactive.messaging.Message$$Lambda$988/0x0000000800a4ec40 to class org.apache.kafka.common.serialization.DoubleSerializer specified in value.serializer
Caused by: java.lang.ClassCastException: class org.eclipse.microprofile.reactive.messaging.Message$$Lambda$988/0x0000000800a4ec40 cannot be cast to class java.lang.Double (org.eclipse.microprofile.reactive.messaging.Message$$Lambda$988/0x0000000800a4ec40 is in unnamed module of loader java.net.URLClassLoader @f5167f0; java.lang.Double is in module java.base of loader 'bootstrap')
	at org.apache.kafka.common.serialization.DoubleSerializer.serialize(DoubleSerializer.java:19)
	at org.apache.kafka.common.serialization.Serializer.serialize(Serializer.java:62)
	at org.apache.kafka.clients.producer.KafkaProducer.doSend(KafkaProducer.java:925)
	at org.apache.kafka.clients.producer.KafkaProducer.send(KafkaProducer.java:885)
	at io.smallrye.reactive.messaging.kafka.impl.ReactiveKafkaProducer.lambda$send$3(ReactiveKafkaProducer.java:102)
	at io.smallrye.mutiny.operators.uni.builders.UniCreateWithEmitter.subscribe(UniCreateWithEmitter.java:22)
	at io.smallrye.mutiny.operators.AbstractUni.subscribe(AbstractUni.java:36)
	at io.smallrye.mutiny.operators.uni.UniRunSubscribeOn.lambda$subscribe$0(UniRunSubscribeOn.java:27)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```
So essentially it seems to be wanting to serialise the whole Message with the value serializer, rather than just the payload when doing it this way.

I tried a few other things and they work fine (the source method is the same as in the previous example). e.g:
```
    @Incoming("one")
    @Outgoing("prices-out")
    public Message<Double> generate(Double d) {
        return Message.of(d);
    }
```
Also this slightly pointless one
```
   @Incoming("one")
    @Outgoing("two")
    public Message<Double> generate(Double d) {
        return Message.of(d);
    }

    @Incoming("two")
    @Outgoing("prices-out")
    public Publisher<Message<Double>> generate(Publisher<Message<Double>> messages) {
        return messages;
    }
```

It fails again if I put the RSO stuff in an earlier method
```
    @Incoming("one")
    @Outgoing("two")
    public Publisher<Message<Double>> generate(Publisher<Double> messages) {
        return ReactiveStreams.fromPublisher(messages)
                .map(d -> Message.of(d))
                .buildRs();
    }

    @Incoming("two")
    @Outgoing("prices-out")
    public Message<Double> generate(Message<Double> m) {
        return m;
    }
```

If I change this last example to 
```
    @Incoming("one")
    @Outgoing("two")
    public Publisher<Message<Double>> generate(Publisher<Double> messages) {
        return ReactiveStreams.fromPublisher(messages)
                .map(d -> Message.of(d))
                .buildRs();
    }

    @Incoming("two")
    @Outgoing("prices-out")
    public Message<Double> generate(Message<Double> m) {
        return Message.of(m());
    }
```
Then, I get the classcast exception when calling the Message.of() method.

Poking a bit further in the last method, I can see that the message is double wrapped.
```
    @Incoming("two")
    @Outgoing("prices-out")
    public Message<Double> generate(Message<Double> m) {
        System.out.println("---> " + m);
        System.out.println("---> " + m.getPayload());
        Message<Double> m2 = (Message) ((Object) m.getPayload());
        System.out.println("---> " + m2.getPayload());
        return m2;
    }
Output:
---> org.eclipse.microprofile.reactive.messaging.Message$$Lambda$969/0x0000000800991040@74dbf530
---> org.eclipse.microprofile.reactive.messaging.Message$$Lambda$969/0x0000000800991040@71e529ed
---> 0.2821780018692911
```
I double checked the previous method, to see if what it receives is actually just the Double payload and it is
```
    @Incoming("one")
    @Outgoing("two")
    public Publisher<Message<Double>> generate(Publisher<Double> messages) {
        return ReactiveStreams.fromPublisher(messages)
                .map(d -> {
                    System.out.println("!!!!! " + d);
                    return Message.of(d);
                })
                .buildRs();
    }

    @Incoming("two")
    @Outgoing("prices-out")
    public Message<Double> generate(Message<Double> m) {
        System.out.println("---> " + m);
        System.out.println("---> " + m.getPayload());
        Message<Double> m2 = (Message) ((Object) m.getPayload());
        System.out.println("---> " + m2.getPayload());
        return m2;
    }
Output:
!!!!! 0.996955166651525
---> org.eclipse.microprofile.reactive.messaging.Message$$Lambda$600/0x0000000800725040@1e22484b
---> org.eclipse.microprofile.reactive.messaging.Message$$Lambda$600/0x0000000800725040@1d069ff
---> 0.996955166651525
```
So it seems that somehow when using RSO to process the streams and create Message instances that these get double-wrapped.

--- a/smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/ProcessorMediator.java
+++ b/smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/ProcessorMediator.java
@@ -1,5 +1,7 @@
 package io.smallrye.reactive.messaging;
 
+import static io.smallrye.reactive.messaging.MediatorConfiguration.Production.STREAM_OF_MESSAGE;
+import static io.smallrye.reactive.messaging.MediatorConfiguration.Production.STREAM_OF_PAYLOAD;
 import static io.smallrye.reactive.messaging.i18n.ProviderExceptions.ex;
 import static io.smallrye.reactive.messaging.i18n.ProviderMessages.msg;
 
@@ -29,6 +31,17 @@ public class ProcessorMediator extends AbstractMediator {
         super(configuration);
         if (configuration.shape() != Shape.PROCESSOR) {
             throw ex.illegalArgumentForProcessorShape(configuration.shape());
+        }
+
+        // IMPORTANT When returning a Multi, Publisher or a PublisherBuilder, you can't mix payloads and messages.
+        if (configuration.production() == STREAM_OF_MESSAGE
+                && configuration.consumption() == MediatorConfiguration.Consumption.PAYLOAD) {
+            throw ex.definitionProduceMessageStreamAndConsumePayload(configuration.methodAsString());
+        }
+
+        if (configuration.production() == STREAM_OF_PAYLOAD
+                && configuration.consumption() == MediatorConfiguration.Consumption.MESSAGE) {
+            throw ex.definitionProducePayloadStreamAndConsumeMessage(configuration.methodAsString());
         }
     }
 
@@ -72,6 +85,7 @@ public class ProcessorMediator extends AbstractMediator {
         // 11. CompletionStage<O> method(I payload)
         // 12. CompletionStage<Message<O>> method(Message<I> msg)
 
+        // IMPORTANT When returning a Publisher or a PublisherBuilder, you can't mix payloads and messages
         switch (configuration.production()) {
             case STREAM_OF_MESSAGE:
                 // Case 1, 3, 5, 7

--- a/smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/StreamTransformerMediator.java
+++ b/smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/StreamTransformerMediator.java
@@ -23,6 +23,17 @@ public class StreamTransformerMediator extends AbstractMediator {
 
     public StreamTransformerMediator(MediatorConfiguration configuration) {
         super(configuration);
+
+        // We can't mix payloads and messages
+        if (configuration.consumption() == MediatorConfiguration.Consumption.STREAM_OF_MESSAGE
+                && configuration.production() == MediatorConfiguration.Production.STREAM_OF_PAYLOAD) {
+            throw ex.definitionProducePayloadStreamAndConsumeMessageStream(configuration.methodAsString());
+        }
+
+        if (configuration.consumption() == MediatorConfiguration.Consumption.STREAM_OF_PAYLOAD
+                && configuration.production() == MediatorConfiguration.Production.STREAM_OF_MESSAGE) {
+            throw ex.definitionProduceMessageStreamAndConsumePayloadStream(configuration.methodAsString());
+        }
     }
 
     @Override

--- a/smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/i18n/ProviderExceptions.java
+++ b/smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/i18n/ProviderExceptions.java
@@ -262,4 +262,15 @@ public interface ProviderExceptions {
             + "The method must return `void`, found %s.")
     DefinitionException definitionReturnVoid(String methodAsString, String returnType);
 
+    @Message(id = 82, value = "Invalid method signature for %s - method returning a Multi<Message<O>>, Publisher<Message<O>> or a PublisherBuilder<Message<O>> cannot consume an individual payload. You must consume a Message<I> instead and handle the acknowledgement.")
+    DefinitionException definitionProduceMessageStreamAndConsumePayload(String methodAsString);
+
+    @Message(id = 83, value = "Invalid method signature for %s - method returning a Multi<O>, Publisher<O> or a PublisherBuilder<O> cannot consume an individual message. You must consume a payload instead or return a Publisher<Message<O>> or a PublisherBuilder<Message<O>>.")
+    DefinitionException definitionProducePayloadStreamAndConsumeMessage(String methodAsString);
+
+    @Message(id = 84, value = "Invalid method signature for %s - method returning a Multi<O>, Publisher<O> or a PublisherBuilder<O> cannot consume a Publisher<Message<I>> or a Multi<Message<I>>. You must consume a Multi<I> or a Publisher<I> instead.")
+    DefinitionException definitionProducePayloadStreamAndConsumeMessageStream(String messageAsString);
+
+    @Message(id = 85, value = "Invalid method signature for %s - method returning a Multi<Message<O>>, Publisher<Message<O>> or a PublisherBuilder<Message<O>> cannot consume a Publisher<I> or a Multi<I>. You must consume a Multi<Message<I>> or a Publisher<Message<I>> instead and handle the acknowledgement.")
+    DefinitionException definitionProduceMessageStreamAndConsumePayloadStream(String messageAsString);
 }

--- a/smallrye-reactive-messaging-provider/src/test/java/io/smallrye/reactive/messaging/invalid/InvalidProcessorSignatureTest.java
+++ b/smallrye-reactive-messaging-provider/src/test/java/io/smallrye/reactive/messaging/invalid/InvalidProcessorSignatureTest.java
@@ -1,0 +1,265 @@
+package io.smallrye.reactive.messaging.invalid;
+
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.awaitility.Awaitility.await;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.spi.DeploymentException;
+
+import org.eclipse.microprofile.reactive.messaging.Incoming;
+import org.eclipse.microprofile.reactive.messaging.Message;
+import org.eclipse.microprofile.reactive.messaging.Outgoing;
+import org.eclipse.microprofile.reactive.streams.operators.PublisherBuilder;
+import org.eclipse.microprofile.reactive.streams.operators.ReactiveStreams;
+import org.junit.jupiter.api.Test;
+import org.reactivestreams.Publisher;
+
+import io.smallrye.mutiny.Multi;
+import io.smallrye.reactive.messaging.WeldTestBaseWithoutTails;
+
+public class InvalidProcessorSignatureTest extends WeldTestBaseWithoutTails {
+
+    @Test
+    public void testPayloadToPayload() {
+        addBeanClass(Tails.class);
+        addBeanClass(PayloadToPayload.class);
+        this.initialize();
+
+        Tails bean = get(Tails.class);
+        await().until(() -> bean.list().size() == 3);
+
+    }
+
+    @Test
+    public void testPayloadToMessage() {
+        addBeanClass(Tails.class);
+        addBeanClass(PayloadToMessage.class);
+        this.initialize();
+
+        Tails bean = get(Tails.class);
+        await().until(() -> bean.list().size() == 3);
+    }
+
+    @Test
+    public void testMessageToMessage() {
+        addBeanClass(Tails.class);
+        addBeanClass(MessageToMessage.class);
+        this.initialize();
+
+        Tails bean = get(Tails.class);
+        await().until(() -> bean.list().size() == 3);
+    }
+
+    @Test
+    public void testMessageToPayload() {
+        addBeanClass(Tails.class);
+        addBeanClass(MessageToPayload.class);
+        this.initialize();
+
+        Tails bean = get(Tails.class);
+        await().until(() -> bean.list().size() == 3);
+    }
+
+    @Test
+    public void testPayloadToPayloads() {
+        addBeanClass(Tails.class);
+        addBeanClass(PayloadToPayloads.class);
+        this.initialize();
+
+        Tails bean = get(Tails.class);
+        await().until(() -> bean.list().size() == 3);
+    }
+
+    @Test
+    public void testPayloadToMessages() {
+        addBeanClass(Tails.class);
+        addBeanClass(PayloadToMessages.class);
+        assertThatThrownBy(this::initialize)
+                .isInstanceOf(DeploymentException.class);
+    }
+
+    @Test
+    public void testMessageToMessages() {
+        addBeanClass(Tails.class);
+        addBeanClass(MessageToMessages.class);
+        this.initialize();
+
+        Tails bean = get(Tails.class);
+        await().until(() -> bean.list().size() == 3);
+    }
+
+    @Test
+    public void testMessageToPayloads() {
+        addBeanClass(Tails.class);
+        addBeanClass(MessageToPayloads.class);
+        assertThatThrownBy(this::initialize)
+                .isInstanceOf(DeploymentException.class);
+    }
+
+    @Test
+    public void testPayloadToPayloadsBuilder() {
+        addBeanClass(Tails.class);
+        addBeanClass(PayloadToPayloadsBuilder.class);
+        this.initialize();
+
+        Tails bean = get(Tails.class);
+        await().until(() -> bean.list().size() == 3);
+    }
+
+    @Test
+    public void testPayloadToMessagesBuilder() {
+        addBeanClass(Tails.class);
+        addBeanClass(PayloadToMessagesBuilder.class);
+        assertThatThrownBy(this::initialize)
+                .isInstanceOf(DeploymentException.class);
+    }
+
+    @Test
+    public void testMessageToMessagesBuilder() {
+        addBeanClass(Tails.class);
+        addBeanClass(MessageToMessagesBuilder.class);
+        this.initialize();
+
+        Tails bean = get(Tails.class);
+        await().until(() -> bean.list().size() == 3);
+    }
+
+    @Test
+    public void testMessageToPayloadsBuilder() {
+        addBeanClass(Tails.class);
+        addBeanClass(MessageToPayloadsBuilder.class);
+        assertThatThrownBy(this::initialize)
+                .isInstanceOf(DeploymentException.class);
+    }
+
+    @ApplicationScoped
+    public static class PayloadToPayload {
+        @Incoming("i")
+        @Outgoing("o")
+        public String consume(String i) {
+            return i;
+        }
+    }
+
+    @ApplicationScoped
+    public static class PayloadToMessage {
+        @Incoming("i")
+        @Outgoing("o")
+        public Message<String> consume(String i) {
+            return Message.of(i);
+        }
+    }
+
+    @ApplicationScoped
+    public static class MessageToMessage {
+        @Incoming("i")
+        @Outgoing("o")
+        public Message<String> consume(Message<String> i) {
+            return i.withPayload(i.getPayload().toLowerCase());
+        }
+    }
+
+    @ApplicationScoped
+    public static class MessageToPayload {
+        @Incoming("i")
+        @Outgoing("o")
+        public String consume(Message<String> i) {
+            return i.getPayload();
+        }
+    }
+
+    @ApplicationScoped
+    public static class PayloadToPayloads {
+        @Incoming("i")
+        @Outgoing("o")
+        public Publisher<String> consume(String i) {
+            return Multi.createFrom().item(i);
+        }
+    }
+
+    @ApplicationScoped
+    public static class PayloadToPayloadsBuilder {
+        @Incoming("i")
+        @Outgoing("o")
+        public PublisherBuilder<String> consume(String i) {
+            return ReactiveStreams.of(i);
+        }
+    }
+
+    @ApplicationScoped
+    public static class PayloadToMessages {
+        @Incoming("i")
+        @Outgoing("o")
+        public Publisher<Message<String>> consume(String i) {
+            return Multi.createFrom().item(Message.of(i));
+        }
+    }
+
+    @ApplicationScoped
+    public static class PayloadToMessagesBuilder {
+        @Incoming("i")
+        @Outgoing("o")
+        public PublisherBuilder<Message<String>> consume(String i) {
+            return ReactiveStreams.of(Message.of(i));
+        }
+    }
+
+    @ApplicationScoped
+    public static class MessageToMessages {
+        @Incoming("i")
+        @Outgoing("o")
+        public Publisher<Message<String>> consume(Message<String> i) {
+            return Multi.createFrom().item(i.withPayload(i.getPayload().toLowerCase()));
+        }
+    }
+
+    @ApplicationScoped
+    public static class MessageToMessagesBuilder {
+        @Incoming("i")
+        @Outgoing("o")
+        public PublisherBuilder<Message<String>> consume(Message<String> i) {
+            return ReactiveStreams.of(i.withPayload(i.getPayload().toLowerCase()));
+        }
+    }
+
+    @ApplicationScoped
+    public static class MessageToPayloads {
+        @Incoming("i")
+        @Outgoing("o")
+        public Publisher<String> consume(Message<String> i) {
+            return Multi.createFrom().item(i.getPayload());
+        }
+    }
+
+    @ApplicationScoped
+    public static class MessageToPayloadsBuilder {
+        @Incoming("i")
+        @Outgoing("o")
+        public PublisherBuilder<String> consume(Message<String> i) {
+            return ReactiveStreams.of(i.getPayload());
+        }
+    }
+
+    @ApplicationScoped
+    public static class Tails {
+        @Outgoing("i")
+        public Multi<String> generate() {
+            return Multi.createFrom().items("a", "b", "c");
+        }
+
+        @Incoming("o")
+        public void consume(String s) {
+            list.add(s);
+        }
+
+        private final List<String> list = new CopyOnWriteArrayList<>();
+
+        public List<String> list() {
+            return list;
+        }
+    }
+
+}

--- a/smallrye-reactive-messaging-provider/src/test/java/io/smallrye/reactive/messaging/invalid/InvalidStreamTransformerSignatureTest.java
+++ b/smallrye-reactive-messaging-provider/src/test/java/io/smallrye/reactive/messaging/invalid/InvalidStreamTransformerSignatureTest.java
@@ -1,0 +1,184 @@
+package io.smallrye.reactive.messaging.invalid;
+
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.awaitility.Awaitility.await;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.spi.DeploymentException;
+
+import org.eclipse.microprofile.reactive.messaging.Incoming;
+import org.eclipse.microprofile.reactive.messaging.Message;
+import org.eclipse.microprofile.reactive.messaging.Outgoing;
+import org.eclipse.microprofile.reactive.streams.operators.PublisherBuilder;
+import org.junit.jupiter.api.Test;
+
+import io.smallrye.mutiny.Multi;
+import io.smallrye.reactive.messaging.WeldTestBaseWithoutTails;
+
+public class InvalidStreamTransformerSignatureTest extends WeldTestBaseWithoutTails {
+
+    @Test
+    public void testPayloadToPayload() {
+        addBeanClass(Tails.class);
+        addBeanClass(PayloadToPayload.class);
+        this.initialize();
+
+        Tails bean = get(Tails.class);
+        await().until(() -> bean.list().size() == 3);
+
+    }
+
+    @Test
+    public void testPayloadToMessage() {
+        addBeanClass(Tails.class);
+        addBeanClass(PayloadToMessage.class);
+        assertThatThrownBy(this::initialize).isInstanceOf(DeploymentException.class);
+    }
+
+    @Test
+    public void testMessageToMessage() {
+        addBeanClass(Tails.class);
+        addBeanClass(MessageToMessage.class);
+        this.initialize();
+
+        Tails bean = get(Tails.class);
+        await().until(() -> bean.list().size() == 3);
+    }
+
+    @Test
+    public void testMessageToPayload() {
+        addBeanClass(Tails.class);
+        addBeanClass(MessageToPayload.class);
+        assertThatThrownBy(this::initialize).isInstanceOf(DeploymentException.class);
+    }
+
+    @Test
+    public void testPayloadToPayloadBuilders() {
+        addBeanClass(Tails.class);
+        addBeanClass(PayloadToPayloadBuilders.class);
+        this.initialize();
+
+        Tails bean = get(Tails.class);
+        await().until(() -> bean.list().size() == 3);
+
+    }
+
+    @Test
+    public void testPayloadToMessageBuilders() {
+        addBeanClass(Tails.class);
+        addBeanClass(PayloadToMessageBuilders.class);
+        assertThatThrownBy(this::initialize).isInstanceOf(DeploymentException.class);
+    }
+
+    @Test
+    public void testMessageToMessageBuilders() {
+        addBeanClass(Tails.class);
+        addBeanClass(MessageToMessageBuilders.class);
+        this.initialize();
+
+        Tails bean = get(Tails.class);
+        await().until(() -> bean.list().size() == 3);
+    }
+
+    @Test
+    public void testMessageToPayloadBuilders() {
+        addBeanClass(Tails.class);
+        addBeanClass(MessageToPayloadBuilders.class);
+        assertThatThrownBy(this::initialize).isInstanceOf(DeploymentException.class);
+    }
+
+    @ApplicationScoped
+    public static class PayloadToPayload {
+        @Incoming("i")
+        @Outgoing("o")
+        public Multi<String> consume(Multi<String> i) {
+            return i;
+        }
+    }
+
+    @ApplicationScoped
+    public static class PayloadToPayloadBuilders {
+        @Incoming("i")
+        @Outgoing("o")
+        public PublisherBuilder<String> consume(PublisherBuilder<String> i) {
+            return i;
+        }
+    }
+
+    @ApplicationScoped
+    public static class PayloadToMessage {
+        @Incoming("i")
+        @Outgoing("o")
+        public Multi<Message<String>> consume(Multi<String> i) {
+            return i.map(Message::of);
+        }
+    }
+
+    @ApplicationScoped
+    public static class PayloadToMessageBuilders {
+        @Incoming("i")
+        @Outgoing("o")
+        public PublisherBuilder<Message<String>> consume(PublisherBuilder<String> i) {
+            return i.map(Message::of);
+        }
+    }
+
+    @ApplicationScoped
+    public static class MessageToMessage {
+        @Incoming("i")
+        @Outgoing("o")
+        public Multi<Message<String>> consume(Multi<Message<String>> i) {
+            return i.map(m -> m.withPayload(m.getPayload().toLowerCase()));
+        }
+    }
+
+    @ApplicationScoped
+    public static class MessageToMessageBuilders {
+        @Incoming("i")
+        @Outgoing("o")
+        public PublisherBuilder<Message<String>> consume(PublisherBuilder<Message<String>> i) {
+            return i.map(m -> m.withPayload(m.getPayload().toLowerCase()));
+        }
+    }
+
+    @ApplicationScoped
+    public static class MessageToPayload {
+        @Incoming("i")
+        @Outgoing("o")
+        public Multi<String> consume(Multi<Message<String>> i) {
+            return i.map(Message::getPayload);
+        }
+    }
+
+    @ApplicationScoped
+    public static class MessageToPayloadBuilders {
+        @Incoming("i")
+        @Outgoing("o")
+        public PublisherBuilder<String> consume(PublisherBuilder<Message<String>> i) {
+            return i.map(Message::getPayload);
+        }
+    }
+
+    @ApplicationScoped
+    public static class Tails {
+        @Outgoing("i")
+        public Multi<String> generate() {
+            return Multi.createFrom().items("a", "b", "c");
+        }
+
+        @Incoming("o")
+        public void consume(String s) {
+            list.add(s);
+        }
+
+        private final List<String> list = new CopyOnWriteArrayList<>();
+
+        public List<String> list() {
+            return list;
+        }
+    }
+
+}

